src/anyp/Uri.cc

```diff
@@ -81,6 +81,66 @@ AnyP::Uri::Encode(const SBuf &buf, const CharacterSet &ignore)
     return output;
 }
 
+SBuf
+AnyP::Uri::Rfc3986Encode(const SBuf &buf)
+{
+    return Encode(buf, CharacterSet::RFC3986_UNRESERVED);
+}
+
+SBuf
+AnyP::Uri::Decode(const SBuf &buf)
+{
+    if (buf.isEmpty())
+        return buf;
+
+    SBuf output;
+    output.reserveSpace(buf.length()); // worst case: buf is not encoded
+
+    const auto end = buf.end();
+    for (auto i = buf.begin(); i != end; ++i) {
+        const auto ch = *i;
+        if (ch != '%') {
+            output.append(ch);
+            continue;
+        }
+        // ch is '%', let's look ahead
+        ++i;
+        if (i == end) {
+            // '%' is the last character, so it's not encoded
+            output.append(ch);
+            break;
+        }
+        const auto ch1 = *i;
+        if (ch1 == '%') {
+            // met "%%"; output "%" and continue
```

Comment:
As this is a parser, I chose to err on the side of accepting input which might be invalid.
If there is consensus, I can remove that bit; @yadij what do you think?

Reply:
> As this is a parser, I chose to err on the side of accepting input which might be invalid.

A parser that "accepts" invalid input needs a very good justification for doing so. "Being a parser" is not a justification (arguably, "being a parser" justifies the _opposite_ behavior!). A parser that, besides accepting invalid input, produces one of several possible output variations from that invalid input also needs to justify the chosen variation.

These basic principles are especially important when parser output is meant, in part, for external consumption (e.g., a decoded and re-encoded request URL is then forwarded to a server) because all external consumers lose the ability to detect previously invalid input (triggering the usual set of "hops A and B disagree on what the URL is" vulnerabilities) and because some external consumers may rely on the original/invalid input (i.e. the proxy is perceived as the one breaking things that "work" without it).

> If there is consensus, I can remove that bit; @yadij what do you think?

FWIW, there is certainly no consensus about the current code.

P.S. The above thoughts apply to all cases where the proposed parser is accepting invalid input, of course (`%%` is just one of them).

Reply:
IMO, in this case "Accept" means not decoding the sequence.

Reply:
Changed my mind after considering https://github.com/squid-cache/squid/pull/1626#discussion_r1436560296 . The decoder should throw when invalid `pct-encoded` input is received.

Reply:
> Changed my mind after considering [#1626 (comment)](https://github.com/squid-cache/squid/pull/1626#discussion_r1436560296) . The decoder should throw when invalid `pct-encoded` input is received.

Yeah, that was how my thoughts evolved as well. Good, I see we have consensus that invalid input should be rejected. Alex and I also discussed using std::optional<SBuf> as a possible way to signal errors; I think I like throwing more and for now I'll use that, but the point is not yet settled